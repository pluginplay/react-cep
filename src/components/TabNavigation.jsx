import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import TabNavigationButton from './TabNavigationButton'

class TabNavigation extends React.Component {
  state = {
    tabClickEvents: {}
  }

  constructor (props) {
    super(props)
    this.state.tabClickEvents = this._getTabClickEvents(props.tabs)
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.tabs !== this.props.tabs) {
      this.setState({ tabClickEvents: this._getTabClickEvents(nextProps.tabs) })
    }
  }

  _getTabClickEvents (tabs) {
    return tabs.reduce((events, tab) => {
      events[tab.identifier] = this.onTabClick.bind(this, tab.identifier)
      return events
    }, {})
  }

  onTabClick (identifier) {
    this.props.onTabSelected(identifier)
  }

  render () {
    const tabButtons = this.props.tabs.map((tab) => {
      const isSelected = this.props.selectedTab === tab.identifier
      const image = `icons/${tab.icon}.svg`
      return (
        <TabNavigationButton key={tab.identifier} isSelected={isSelected}
          onClick={this.state.tabClickEvents[tab.identifier]}>
          <img src={image} alt={tab.title} />
        </TabNavigationButton>
      )
    })
    return (
      <div className={this.props.className}>
        {tabButtons}
      </div>
    )
  }
}

TabNavigation.propTypes = {
  tabs: PropTypes.arrayOf(PropTypes.object).isRequired,
  selectedTab: PropTypes.string.isRequired,
  onTabSelected: PropTypes.func.isRequired
}

export default styled(TabNavigation)`
  margin: 0 0 0 auto;
  background: ${props => props.theme.background.dark};
  width: 35px;
`
