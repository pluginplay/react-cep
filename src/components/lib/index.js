import SystemState from '../../models/SystemState'
import PersistedState from '../../models/PersistedState'

export { SystemState, PersistedState }
