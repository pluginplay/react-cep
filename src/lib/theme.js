import * as csInterface from 'cep-lib/csinterface'
import { lighten, darken } from 'polished'

export const getDefaultTheme = () => {
  const ae = csInterface.getThemeInformation()
  ae.systemHighlightColor = '#0078d7' // We have to hard-code this because Windows gets the wrong color.
  return {
    ae,
    background: {
      lighter: lighten(0.2, ae.panelBackgroundColor),
      light: lighten(0.1, ae.panelBackgroundColor),
      dark: darken(0.05, ae.panelBackgroundColor),
      darker: darken(0.1, ae.panelBackgroundColor)
    },
    colors: {
      green: '#23EB87',
      red: '#eb093f',
      blue: ae.systemHighlightColor
    }
  }
}
