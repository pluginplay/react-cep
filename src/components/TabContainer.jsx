import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import TabNavigation from './TabNavigation'
import Flex from 'styled-flex-component'

class TabContainer extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      selectedTab: this._getDefaultTab(props)
    }
    this._onTabSelectedBound = this.onTabSelected.bind(this)
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.tabs !== this.props.tabs) {
      this.setState({ selectedTab: this._getDefaultTab(nextProps) })
    }
  }

  _getDefaultTab (props) {
    return props.tabs[0].identifier
  }

  onTabSelected (tabIdentifier) {
    this.setState({ selectedTab: tabIdentifier })
  }

  renderContent () {
    const selectedTab = this.props.tabs.filter((tab) => tab.identifier === this.state.selectedTab)[0]
    if (selectedTab) {
      const TabComponent = selectedTab.component
      return (<TabComponent />)
    }
  }

  render () {
    return (
      <Flex full className={this.props.className}>
        <div className={'content'}>
          {this.renderContent()}
        </div>
        <TabNavigation tabs={this.props.tabs} selectedTab={this.state.selectedTab}
          onTabSelected={this._onTabSelectedBound} />
      </Flex>
    )
  }
}

TabContainer.propTypes = {
  tabs: PropTypes.arrayOf(PropTypes.object).isRequired
}

export default styled(TabContainer)`
  position: absolute;
  top: 0;
  bottom: 0;
  .content {
    margin: 0 auto 0 0;
    padding: 10px;
    width: 100%;
    overflow-y: auto;
    position: relative;
  }
  @media (max-width: 320) {
    .content { padding: 3px; }
  }
`
