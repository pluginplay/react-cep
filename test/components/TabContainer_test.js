import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { expect, withTheme } from '../setup'
import { getDefaultTheme } from '../../src/lib/theme'
import TabContainer from '../../src/components/TabContainer'

import TabNavigation from '../../src/components/TabNavigation'

let tabs = null

class TabA extends React.Component { render () { return <div>Tab A</div> } }
class TabB extends React.Component { render () { return <div>Tab B</div> } }
class TabC extends React.Component { render () { return <div>Tab C</div> } }

beforeEach(() => {
  tabs = [
    { identifier: 'tab-1', title: 'Tab 1', icon: 'one', component: TabA },
    { identifier: 'tab-2', title: 'Tab 2', icon: 'two', component: TabB },
    { identifier: 'tab-3', title: 'Tab 3', icon: 'three', component: TabC }
  ]
})

const getWrapper = (props = {}) => {
  return shallow(<TabContainer theme={getDefaultTheme()} tabs={tabs} {...props} />).dive()
}

it('renders correctly', () => {
  const tree = renderer
    .create(withTheme(<TabContainer tabs={tabs} />))
    .toJSON()
  jestExpect(tree).toMatchSnapshot()
})

it('sets the default tab to the first tab', () => {
  const wrapper = getWrapper()
  expect(wrapper.find(TabA)).to.have.length(1)
  expect(wrapper.find(TabB)).to.have.length(0)
  expect(wrapper.find(TabC)).to.have.length(0)
  expect(wrapper.find(TabNavigation).first()).to.have.prop('selectedTab').equal('tab-1')
})

context('second tab selected', () => {
  let wrapper = null
  beforeEach(() => {
    wrapper = getWrapper()
    wrapper.find(TabNavigation).first().simulate('tabSelected', 'tab-2')
  })

  it('marks the second tab as selected in TabNavigation', () => {
    expect(wrapper.find(TabNavigation).first()).to.have.prop('selectedTab').equal('tab-2')
  })

  it('renders the component corresponding with the second tab', () => {
    expect(wrapper.find(TabA)).to.have.length(0)
    expect(wrapper.find(TabB)).to.have.length(1)
    expect(wrapper.find(TabC)).to.have.length(0)
  })
})

