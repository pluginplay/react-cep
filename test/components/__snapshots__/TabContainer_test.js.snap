// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #262626;
  cursor: pointer;
  -webkit-transition: background 0.1s linear;
  transition: background 0.1s linear;
  width: 100%;
  height: 35px;
  border: none;
  outline: none;
}

.c3 img {
  width: 16px;
  -webkit-transition: -webkit-transform 0.25s ease-in-out;
  -webkit-transition: transform 0.25s ease-in-out;
  transition: transform 0.25s ease-in-out;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  margin: 0 auto;
}

.c3:hover {
  background: #262626;
}

.c3:active {
  padding: 0;
}

.c3:active img {
  -webkit-transform: scale(0.9);
  -ms-transform: scale(0.9);
  transform: scale(0.9);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #191919;
  cursor: pointer;
  -webkit-transition: background 0.1s linear;
  transition: background 0.1s linear;
  width: 100%;
  height: 35px;
  border: none;
  outline: none;
}

.c4 img {
  width: 16px;
  -webkit-transition: -webkit-transform 0.25s ease-in-out;
  -webkit-transition: transform 0.25s ease-in-out;
  transition: transform 0.25s ease-in-out;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  margin: 0 auto;
}

.c4:hover {
  background: #262626;
}

.c4:active {
  padding: 0;
}

.c4:active img {
  -webkit-transform: scale(0.9);
  -ms-transform: scale(0.9);
  transform: scale(0.9);
}

.c2 {
  margin: 0 0 0 auto;
  background: #191919;
  width: 35px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  height: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
}

.c0 {
  position: absolute;
  top: 0;
  bottom: 0;
}

.c0 .content {
  margin: 0 auto 0 0;
  padding: 10px;
  width: 100%;
  overflow-y: auto;
  position: relative;
}

<div
  className="c0 c1"
>
  <div
    className="content"
  >
    <div>
      Tab A
    </div>
  </div>
  <div
    className="c2"
  >
    <button
      className="c3"
      onClick={[Function]}
    >
      <img
        alt="Tab 1"
        src="one"
      />
    </button>
    <button
      className="c4"
      onClick={[Function]}
    >
      <img
        alt="Tab 2"
        src="two"
      />
    </button>
    <button
      className="c4"
      onClick={[Function]}
    >
      <img
        alt="Tab 3"
        src="three"
      />
    </button>
  </div>
</div>
`;
